### ISSUE TITLE
Question: suggested best practice for holding onto a store constructed with props

### ISSUE THREAD

**Issue by damonmaria** [mobxjs/mobx-react-lite#74]
First up: we've switched a project over from `mobx-react` to hooks and `mobx-react-lite` and are loving it. Thanks.

Second: apologies for a question as a Github issue. 

A common pattern we have is a component wanting to hold onto a mobx store only for the component's lifetime, but the store constructor is passed a prop and we want to re-create the store if the prop changes. 

So far I'm using `useMemo` to manage this. For example, with a `react-router` param passed in:
```typescript
const DashboardPage: FC<RouteComponentProps<IRouteParams>> = ({ match }) => {
  const { siteId } = match.params
  const store = useMemo(() => new DashboardStore(siteId), [siteId])
```

While this works fine the [React reference for `useMemo`](https://reactjs.org/docs/hooks-reference.html#usememo) has a warning: 

> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.

The store being recreated would be bad, the user losing data. I'm assuming this is a common use case not just for me. What is the 'mobx recommended' way to handle this?

**Comment by FredyC** [comment]
Yea, `useMemo` is definitely a bad candidate for that. What about `useRef` instead? It does not have a lazy init capability, but it's easy to make a custom hook for that.

```js
function useStore(storeFactory) {
  const storeRef = React.useRef()
  if (!storeRef.current) {
    storeRef.current = storeFactory()
  }
  return storeRef.current
}
```

**Comment by urugator** [comment]
```javascript
function useDashboardStore(siteId) {
  const [store, setStore] = React.useState(new DashboardStore(siteId));  
  
  useEffect(() => {
    setStore(new DashboardStore(siteId))
  }, [siteId])
    
  return store;
}
```

**Comment by FredyC** [comment]
@urugator That makes total sense! My approach definitely has a weakness if `siteId` from the route would change without remounting, it would not work. So yea, it's surely safer although it makes it harder to abstract as a custom hook. With many similar stores, it would be somewhat non-DRY.

@damonmaria Closing, feel free to open if you have further questions.

**Comment by urugator** [comment]
```javascript
function useStore(factory, deps) {
  const [store, setStore] = React.useState(factory);
  const [isInitial, setIsInitial] = React.useState(true); 
  
  React.useEffect(() => {
    if (!isInitial) {
      setStore(factory())      
    }
    setIsInitial(false)
  }, deps)
    
  return store;
}

// usage
useStore(() => new DashboardStore(siteId), [siteId]);
```

**Comment by FredyC** [comment]
@urugator Nice idea. I guess I am still kinda afraid to overuse `useState` for some reason :) Although for tracking that initial, the `useRef` might suffice since it's just to track a value once. Ultimately it doesn't matter as React won't re-render if a value is unchanged (always setting it to `false`).

I would also note, that if the store would depend on observable value, it could make sense to do this instead. Haven't tested it, but should work.

```js
  useDisposable(() => mobx.autorun(() => {
    if (!isInitial) {
      setStore(factory())      
    }
    setIsInitial(false)
  }), deps)
```

**Comment by damonmaria** [comment]
A slight issue I have with using `useEffect` is that if the prop changes then for one render the prop value and the store will be out of sync (since the store was created from the previous value). While probably not often an issue it could lead to confusing bugs in the render. 

My thought was something along the lines of:
```typescript
export const useStore = <T>(factory: () => T, deps: any[] = []): T => {
  const prevDepsRef = useRef<any[]>([])
  const storeRef = useRef<T | null>(null)
  if (!storeRef.current || !shallowEqual(prevDepsRef.current, deps)) {
    storeRef.current = factory()
    prevDepsRef.current = deps
  } 
  return storeRef.current
}
```

While technically mutating during a render it's similar to how `useMemo` works, and similar to how the React reference talks about using `useRef` for lazy initialization.

**Comment by FredyC** [comment]
Good point @damonmaria. Might be slightly more tricky if your store would depend on some observable value. You would probably need to pass those into `deps` as well, otherwise, it wouldn't know when to throw away the old store.

**Comment by urugator** [comment]
You're right @damonmaria. When I think about it, isn't it a use case for `getDerivedStateFromProps`?
A `useMemo` impl with guaranteed memoization seems more robust and easier to follow.
Perhaps you could call it `useDerived` or something like that...

@FredyC Maybe he could implement `useComputed` which uses `useDerived` instead of `useMemo`...

**Comment by damonmaria** [comment]
Yes, `useMemo` with a guarantee would be perfect. 

I'm not suggesting what I have as the solution but since this isn't a trivial problem to solve, maybe `mobx-react-lite` should provide some form of `useStore` that handles this? Or maybe a `mobx-react-lite-utils` :)

**Comment by FredyC** [comment]
@damonmaria I don't think it's very common to construct store dynamically like that. Or at least I haven't seen it much. What about having a single instance in the ref (or context) and just update it whenever values change? That would make things much easier imo.

**Comment by alsotang** [comment]
Not every use case should be a custom hook. If the case is rare, just use React native hooks.

**Comment by FredyC** [comment]
> Not every use case should be a custom hook. If the case is rare, just use React native hooks.

Some interesting read for you and I agree with it: https://frontarm.com/james-k-nelson/react-hooks-intuition/

Besides, it's not the point really. Extracting to a custom hook does change the task at hand. Perhaps very slightly as it does not need to be universal, but splitting logic generally is always a good thing as it makes you think about the problem more deeply than just "I need to do this".

**Comment by damonmaria** [comment]
@FredyC 
> I don't think it's very common to construct store dynamically like that

Strange. Maybe I'm doing something wrong. Fairly commonly I have stores who's lifetime should match that of a component tree, and often is based off the route params at that point. Outside of that they just don't have the data they need to operate. Especially in the situation of the store being based off route params it has to be coordinated with a component as (react-router) doesn't expose route params apart from through props. 

Another situation where I use this is where there's a dynamic number of the same component and each needs it's own state (state beyond something simple that would suit `useState`). 

How do others handle these situations?

**Comment by alsotang** [comment]
@damonmaria Perhaps I understand the situation you faced.
Recently I write some Vue.js code and face the same problem.

My problem is. I have a complex Editor component. It read `projectId` from url params, and render the whole page base on it. The editor has many internal states.

When I navigate to another project, the url changes so the projectId changes, but Vue will re-use the Editor component. That makes some internal states inconsistent. So my approach is that recreat the editor whenever the url changes by using `key` prop ( https://reactjs.org/docs/lists-and-keys.html#keys     https://vuejs.org/v2/guide/list.html#key).

**Comment by damonmaria** [comment]
@alsotang I noticed in the React docs it mentioning changing the `key` property to achieve that. The thing I don't like about that technique is it's the user of the component that has to know to set the `key` to the dependencies that change, rather than the component itself.

**Comment by alsotang** [comment]
@damonmaria The use of the component doesn't need to set the key. You can just wrap the existing component in a simple HOC. The HOC deal with the `key` prop

**Comment by alsotang** [comment]
@FredyC I think I can understand the article and I agree with it. 
I re-express my words: Not every use case of mobx-react-lite should be a custom hook api of `mobx-react-lite`. Library apis should be universal

**Comment by mweststrate** [comment]
@urugator you should pubish that `useStore` hook on npm, I need that pattern quite regularly :')

**Comment by urugator** [comment]
Complete impl (using memoization):
https://gist.github.com/urugator/5c78da03a7b1a7682919cc1cf68ff8e9
Feel free copy/modify/share/publish/whatever

**Comment by FredyC** [comment]
For the sake of closure here, new documentation includes relevant information.

https://mobx-react.netlify.com/state-outsourcing

Might add a specific recipe later when #159 is merged

**Comment by trickreich** [comment]
This thread is really helpful when coming from "mobx with class components" to "mobx lite with functional components and hooks".

But to be honest: I don't get it...

You are writing your stores still as classes with `computed, observable, autorun, etc..`?

In the docs following is referenced:
`const [bigState] = React.useState(createExpensiveStore)`
But the `createExpensiveStore` method isn't shown at all.
Maybe you can point in some direction when a store is "expensive"?

**Comment by FredyC** [comment]
Have you seen https://mobx-react.netlify.com/state-local?

**Comment by trickreich** [comment]
Yep, read the whole documentation a few times.

**Comment by urugator** [comment]
@trickreich 
> Maybe you can point in some direction when a store is "expensive"?

The `React.useState(something)` runs every render, so you just want to avoid recreating `something` complex every render, simple as that, eg:
```javascript 
// object passed to use state is created every render
// but used only on the initial render
// it's created and trown away on subsequent renders - potentially expensive
const state = useState({ a: "a", b: "b", c: "c"});
```

**Comment by trickreich** [comment]
I know.. Maybe a example is better:
https://codesandbox.io/s/trickreich-mobx-react-lite-testing-lc7nt

**Comment by ivandotv** [comment]
I'm following this discussion, and would like to add my opinion, which might not be directly related to the issue but someone could find it useful.
This article from Michel Weststrate was eye opening for me:
[How to decouple state and UI \(a.k.a. you don’t need componentWillMount\)](https://hackernoon.com/how-to-decouple-state-and-ui-a-k-a-you-dont-need-componentwillmount-cc90b787aa37)
I was experimenting with taking this approach with my applications, and I'm currently in the process of creating a note taking web application using the article as guidance.

I still haven't found a reason to create mobx stores inside the components (I'm using react with hooks exclusively)
All my components are **really dumb**.
My whole application exists without the ui (business logic at least), so that means that I could just open the console and add new notes, tasks... basically drive the whole application from the command line.
I'm also not using `React.context()` for passing down stores (I might be wrong about this, it would be nice if someone could comment on this).

This is an example of the component.
```javascript
export const ListHolder = observer(function(props) {
  const listStore = serviceLocator.get(STORES.LIST_STORE)
  return (
    <div>
      <div className={styles.wrap}>
        {listStore.lists.map(list => {
          return <List key={list.id} list={list} />
        })}
      </div>
    </div>
  )
})


```
The case against the `context` is that my application is **driven** by MobX. If I put my store in react context and pass it to components that way there is no real benefit because that context will never call render on components.
So as you can see on this line: ` const listStore = serviceLocator.get(STORES.LIST_STORE)`
I'm getting the store from outside of the react (using service locator pattern)
`STORES.LIST_STORE` is just a object with with the names of all the stores that are registered with the `serviceLocator`

Testing the component is also easy because just before I instantiate the component for the test I can change the store (use mock or whatever)  that is retrieved with the service locator.

And also when you think about it `React.useContext()` is a kind of a `service locator` except you also need to do all that `provider && consumer` dance.

For component related state `useState` and `useReducer` is sufficient.

Any thoughts on this? Do you foresee any problems with my setup?

**Comment by urugator** [comment]
@ivandotv
 
> except you also need to do all that provider && consumer dance

You're not avoiding this dance by using a *service locator*, but by using `import`. 
`serviceLocator` should be passed via props/context as any other dependency, which in turn makes it kinda pointless.

**Comment by ivandotv** [comment]
Yeah I thought  to pass the service locator as a property itself. But I don't see the point.
Services that are acquired via `serviceLocator` can easily be mocked in tests.

**Comment by joenoon** [comment]
@ivandotv re serviceLocator/STORES/props/context: This sort of makes the app a singleton.  What I personally like better is to initialize stores and similar near the App root, pass them down in context.  This way you could imagine a scenario where `<App key={userToken} />` completely starts a fresh tree with none of the headache and bookkeeping on external stores if the userToken changes.

**Comment by FredyC** [comment]
To be honest, I don't have any real evidence why is store in a global-ish variable a bad thing. The usual answer is because of tests, but Jest runs each test file in a separate process, so these won't bother each other. As long as you reset that global object properly before each test, there shouldn't be any practical harm if I am not mistaken.

Another popular response is SSR which has something to do with parallelism, but I don't use SSR, so I don't understand it that well.

Are there some other reasons except its _weird_ in the world of modules? Is singleton really that bad pattern?


@joenoon 
> This way you could imagine a scenario where `<App key={userToken} />` completely starts a fresh tree with none of the headache and bookkeeping on external stores if the userToken changes.

I wonder what bookkeeping you are talking about. Token could be as well changed with an imperative method on `serviceLocator` object as long as it's prepared for it. MobX will take care of the necessary rendering.

**Comment by urugator** [comment]
@FredyC 

> Is singleton really that bad pattern?

It depedens whether you use it because you're lazy to pass dependencies around or because you actually want to ensure there is only single instance of certain object per application.    

> Another popular response is SSR which has something to do with parallelism, but I don't use SSR, so I don't understand it that well.

It's quite trivial - the "module based" singletons are app scoped (single store per node app), but you need them request scoped (single store per request).

**Comment by FredyC** [comment]
> It depedens whether you use it because you're lazy to pass dependencies around or because you actually want to ensure there is only single instance of certain object per application.

😆 Good point. I would say it's kinda a mix of both. Usually, the app state should exist only once (in terms of client-side). The ease of use comes as a nice side effect.

Thanks for SSR explanation, makes sense.

**Comment by ivandotv** [comment]
> It depends whether you use it because you're lazy to pass dependencies around or because you actually want to ensure there is only single instance of certain object per application.
> 
I agree singletons are bad.
But how come react context/provider is not essentially a singleton? It doesn't have to be but it allmost **always** is.

One other solution that I can think of is to use `context/provider` to pass in the object that holds all `stores/services` basically to pass in the `serviceLocator` 😐
But that feels as two sides of the same coin. 

Also, isn't the mobx root store always a singleton? 
The only question remains is how to test the root store i.e. to create and pass store mocks.

**Comment by FredyC** [comment]
What I love about Context is the ability to override value easily by rendering a Provider. That came very handy in tests several times because I did not need to bother with any sort of cleanup which might be complicated sometimes. I haven't used this ability for a regular app code yet.

As @urugator said, this approach would certainly be problematic in case of SSR. What I would advice is probably have a simple hook `useServiceLocator` (or whatever) which gives you that imported singleton. In case you ever need to worry about SSR, you can just change that hook to `useContext` internally. Additionally, we don't know yet what obstacles the Strict/Concurrent/Suspense will bring, it's good to be ready imo.

> The only question remains is how to test the root store i.e. to create and pass store mocks.

Not sure why would you test store? Isn't it tested enough by using it in the app code? Well, if you require unit testing, then it's understandable, but it's probably overkill.

**Comment by FredyC** [comment]
Oh, I just realized, since I am using mobx-state-tree to form the app state, it would be very problematic to do a proper cleanup. It provides a reset for props, but I am also using volatiles and those cannot be cleaned. I would be ending up with an incoherent state between tests. That definitely stinks.

If you are building your ServiceLocator with classes or whatever, you will have to provide the ability to clean up everything. That doesn't sound nice on its own to expand app code because of tests. Besides, imagine a scenario where you need to add a new property there, but you forget to include that in a clean up routine. That can introduce flakiness to your tests.

So ultimately, is it really such a big deal to have it the Context and let each test have its own store without need to bother with any sort of reset?

**Comment by mweststrate** [comment]
My 2 cents

> To be honest, I don't have any real evidence why is store in a global-ish variable a bad thing. The usual answer is because of tests, but Jest runs each test file in a separate process, so these won't bother each other. As long as you reset that global object properly before each test, there shouldn't be any practical harm afaik.

Resetting singletons is my in experience way more risky in complicated projects that it inititally seems. Not because resetting itself is hard (in MST it is trivial), but because run away promises easily screw up different tests.

> So as you can see on this line: const listStore = serviceLocator.get(STORES.LIST_STORE)

I typically use context itself as the way to distribute all the stores, in other words, context becomes the service locator. This avoids the risks of singletons, and at the same time still keeps the same decoupling as argued in my original blog post. Also note that that context has an immutable state and all state itself lives in the stores. 

That all being said, this is a quite offtopic thread on an already closed issue, so nobody really benefits from the knowledge in here. @trickreich please don't comment on closed issues like this in the future, just because they relate to your question! Just open a new one and optionally refer to the old closed issue. 

Going to lock the conversation now, as it is a very interesting, but different discussion than the original question. By all means do further respond if you want, just go through the effort to first open a new issue :)
